mmv1/third_party/terraform/data_sources/data_source_sql_database_instances.go

```diff
@@ -0,0 +1,195 @@
+				Description: `Project ID of the project that contains the instances.`,
+			},
+			"database_version": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: `To filter out the database instances which are of the specified database version.`,
+			},
+			"region": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: `To filter out the database instances which are located in this specified region.`,
+			},
+			"zone": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: `To filter out the database instances which are located in this specified zone.`,
+			},
+			"tier": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: `To filter out the database instances based on the machine type.`,
+			},
+			"state": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: `To filter out the database instances based on the current state of the database instance, valid values include : "SQL_INSTANCE_STATE_UNSPECIFIED", "RUNNABLE", "SUSPENDED", "PENDING_DELETE", "PENDING_CREATE", "MAINTENANCE" and "FAILED".`,
+			},
+			"instances": {
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem: &schema.Resource{
+					Schema: datasourceSchemaFromResourceSchema(resourceSqlDatabaseInstance().Schema),
+				},
+			},
+		},
+	}
+}
+
+func dataSourceSqlDatabaseInstancesRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+	userAgent, err := generateUserAgentString(d, config.userAgent)
+	if err != nil {
+		return err
+	}
+	project, err := getProject(d, config)
+	if err != nil {
+		return err
+	}
+	filter := ""
+
+	if v, ok := d.GetOk("database_version"); ok {
+		filter += fmt.Sprintf("databaseVersion:%s", v.(string))
+	}
+	if v, ok := d.GetOk("region"); ok {
+		if filter != "" {
+			filter += " AND "
+		}
+		filter += fmt.Sprintf("region:%s", v.(string))
+	}
+	if v, ok := d.GetOk("zone"); ok {
+		if filter != "" {
+			filter += " AND "
+		}
+		filter += fmt.Sprintf("gceZone:%s", v.(string))
+	}
+	if v, ok := d.GetOk("tier"); ok {
+		if filter != "" {
+			filter += " AND "
+		}
+		filter += fmt.Sprintf("settings.tier:%s", v.(string))
+	}
+	if v, ok := d.GetOk("state"); ok {
+		if filter != "" {
+			filter += " AND "
+		}
+		filter += fmt.Sprintf("state:%s", v.(string))
+	}
+	pageToken := ""
+	databaseInstances := make([]map[string]interface{}, 0)
+	for {
+		var instances *sqladmin.InstancesListResponse
+		err = retryTimeDuration(func() (rerr error) {
+			instances, rerr = config.NewSqlAdminClient(userAgent).Instances.List(project).Filter(filter).PageToken(pageToken).Do()
+			return rerr
+		}, d.Timeout(schema.TimeoutRead), isSqlOperationInProgressError)
+		if err != nil {
+			return err
+		}
+
+		pageInstances := flattenDatasourceGoogleDatabaseInstancesList(instances.Items, project)
+		databaseInstances = append(databaseInstances, pageInstances...)
+
+		pToken := instances.NextPageToken
```

Comment:
nit: Let's directly assign the next page token value to pageToken variable, will reduce the code a bit.

Reply:
Done